Reader: Post Sharing via Calypso Editor

blowery opened the pull request:
Pull out Press This support and use the editor inside Calypso.

To test this, open the Reader, find a post you like, hit the Share button and pick WordPress. You should be moved to the editor, with a sample post filled out.

If the post had an embed, the embed will be embedded at the top of the post. Hopefully.

If the post had a canonical image and no embed, the image will be at the top of the post. The image is hotlinked, which I don't love, but it's what Press This does too. 

The text is pulled from the post's excerpt. Would be nice to pull it from selected text, but little steps.

![sharing](https://cloud.githubusercontent.com/assets/14350/13304920/c35da9c0-db25-11e5-95ef-00190cdfed4a.gif)

blowery commented:
cc @aduth

blowery commented:
cc @gwwar too cause she's good at breaking things

So to test this, open the Reader, find a post you like, hit the Share button and pick WordPress. You should be moved to the editor, with a sample post filled out.

If the post had an embed, the embed will be embedded at the top of the post. Hopefully.

If the post had a canonical image and no embed, the image will be at the top of the post. The image is hotlinked, which I don't love, but it's what Press This does too. 

The text is pulled from the post's excerpt. Would be nice to pull it from selected text, but little steps.

gwwar commented on client/post-editor/controller.js:
```diff
@@ -91,13 +92,21 @@ function maybeRedirect( context, postType, site ) {
-		</p>
-	);
+function getPressThisContent( text, url, title, image, embed ) {
+	let pieces = [];
```
const

aduth replied:
> const

My personal preference is that even if `const` lets me `push`, I still like to see `let` as an indicator that the value will change over time.

gwwar replied:
`const` provides us with a constant reference, not immutability. I was looking for a reference change here due to the `let` and was confused when I didn't find one.

If we really feel strongly on using `const` only for primitive values like strings, booleans and numbers, and `let` on class instances, objects, and arrays this should be updated in our coding guidelines.

blowery replied:
I really wouldn't use `const` to try to indicate immutability. That just leads people down the wrong road w.r.t. what it really does.

gwwar commented on client/post-editor/controller.js:
```diff
@@ -143,7 +152,7 @@ module.exports = {
 				// handle press-this params if applicable
 				if ( context.query.url ) {
-					let pressThisContent = getPressThisContent( context.query.text, context.query.url, context.query.title );
+					let pressThisContent = getPressThisContent( context.query.text, context.query.url, context.query.title, context.query.image, context.query.embed );
```
`const`

gwwar commented on client/post-editor/controller.js:
```diff
@@ -91,13 +92,21 @@ function maybeRedirect( context, postType, site ) {
+		pieces.push( ReactDomServer.renderToStaticMarkup( <blockquote>{ text }</blockquote> ) );
+	}
+
+	pieces.push( ReactDomServer.renderToStaticMarkup( <p>via <a href={ url }>{ title }</a>.</p> ) );
```
Do we need to worry about double encoding?

blowery replied:
I don't think so. the post title that we're using passing along has already been de-entified and stripped of tags. I think.

But if someone else tries to pass random stuff, I really don't want it slipping through.

blowery replied:
Do we need to translate the "via" ?

aduth replied:
>Do we need to translate the "via" ?

Probably.

gwwar commented on client/reader/share/index.jsx:
```diff
@@ -43,27 +46,28 @@ var actionMap = {
 	pressThis: function( post ) {
-		var primarySite = sitesList.getPrimary(), wordpressUrl;
-
+		let primarySite = sitesList.getPrimary();
```
`const`

gwwar commented:
:+1: Nice! Looking good so far. I'll try to break it a bit more tomorrow, but I think it's pretty solid.

blowery commented:
Problems with this flow:
- [ ] Getting back to the stream is difficult on desktop. Not as bad in a browser where you can mostly hit back. Maybe we pass a `backto` param into the editor when sharing and head back there after publish?
- [ ] If I want a different image than the feature, it's a pain to find
- [ ] If I want the image, but not the embed, it's a pain to find

blowery commented:
Throwing a Design Review on this one too, as I'd like to let design poke at it a bit.

cc @jancavan and @shaunandrews

jancavan commented:
While this is working great, the main issue I have with this is it disrupts the flow when you're just consuming content because it takes you to the Editor. I'm also required to type something in so the "Publish" button gets activated before I can share/re-blog it.

alisterscott commented:
Overall works well, good job, I like it :+1: 

**Some Specific Issues Found**

1. As mentioned by @jancavan above: when arriving at the editor you can't publish the post immediately without making some changes as the button is disabled, probably to stop people posting empty posts - do we need to override this?
2. The share only shares a new post for the 'primary' blog which is only configurable (AFAIK) via wp-admin, it doesn't take into account the currently selected site, recent sites or favorite sites, and there's no way to change it once the editor has been opened with the content. My suggestion is to show the site chooser (a la new post) when user has multiple sites
3. Items being shared with embedded soundclouds don't create embedded soundclouds (eg. https://testingreadershares.wordpress.com/2016/02/25/rob-coneybeer-on-nest-and-the-next-big-thing-in-hardware/)
4. Items being shared with embedded videopresses don't create embedded videopresses (eg. https://testingreadershares.wordpress.com/2016/02/25/editor-atd-spellchecking-correcting-a-word-with-atd-editor-iphone-6/ ) but YouTube works :+1: 
5. There's no clear way to navigate back to the reader where you shared from, and if you do it via the toolbar, then your place where you shared from is lost (maybe use local storage to remember the place?)
6. There doesn't seem to be any promotion added to the shared post that shows it was shared via the WordPress.com Reader? Should we add a tag or a badge or some content or something? Good marketing potential?
7. The via link is good as it's directly to the title of the source, should we mention the blog/site/author also?

**Tested:** animated GIFs, long posts, short posts, excerpts, mobile size, sharing shared posts (the matrix?), YouTube embeds

mtias commented:
I agree with @alisterscott that we should look at not relying on primary site (even though it can be set in Calypso from `/me/account`) and maybe offer the chance to select which site like we do when you click "new post" vis site-popover or site-dropdown. We could also leverage `sites.getLastSelectedSite()`, but I think those assumption can hurt the UX since it doesn't give you a choice.

mtias commented:
> Getting back to the stream is difficult on desktop. Not as bad in a browser where you can mostly hit back. Maybe we pass a backto param into the editor when sharing and head back there after publish?

We used to have a "← Back" in the editor that would respect this, but it was changed to "← All Posts" since we didn't have a flow to work with. Now that we do, we should update it to read "← Back to Reader" if you are coming from this specific functionality.

mtias commented:
Probably the best way to handle site selection is to allow press this without a site param to go through a site selection step when hitting the `/post` url.

aduth commented on client/post-editor/controller.js:
```diff
@@ -143,7 +152,7 @@ module.exports = {
 				// handle press-this params if applicable
 				if ( context.query.url ) {
-					let pressThisContent = getPressThisContent( context.query.text, context.query.url, context.query.title );
+					const pressThisContent = getPressThisContent( context.query.text, context.query.url, context.query.title, context.query.image, context.query.embed );
```
Some destructuring could help shorten this up:

```js
const { url, text, title, image, embed } = context.query;
```

aduth replied:
Or just revise `getPressThisContent` to accept `context.query`.

blowery replied:
Yeah, I was thinking about doing the latter, will take care of that.

mtias commented:
For site selection flow see #3560

aduth commented on client/post-editor/controller.js:
```diff
@@ -91,13 +92,21 @@ function maybeRedirect( context, postType, site ) {
+function getPressThisContent( text, url, title, image, embed ) {
+	const pieces = [];
+	if ( image ) {
+		pieces.push( ReactDomServer.renderToStaticMarkup( <p><a href={ url }><img src={ image } /></a></p> ) );
```
Wonder if we need the `<p>` wrapper here if we're also adding plain text separated by double-line break.

blowery replied:
Prolly not since we're surrounded by line breaks.

aduth commented on client/post-editor/controller.js:
```diff
@@ -91,13 +92,21 @@ function maybeRedirect( context, postType, site ) {
+	if ( image ) {
+		pieces.push( ReactDomServer.renderToStaticMarkup( <p><a href={ url }><img src={ image } /></a></p> ) );
+	}
+	if ( isValidUrl( embed ) ) {
```
What's an example of an invalid URL?

blowery replied:
`<script>alert('hi')</script>`

aduth replied:
>`<script>alert('hi')</script>`

Maybe we should leverage `ReactDomServer` then, with added bonus of consistency?

```jsx
pieces.push( ReactDomServer.renderToStaticMarkup( <p>{ embed }</p> ) );
```

aduth commented on client/reader/share/index.jsx:
```diff
@@ -43,27 +46,28 @@ var actionMap = {
+		const args = {};
+
+		if ( post.content_embeds && post.content_embeds.length ) {
+			args.embed = post.content_embeds[0].src;
```
If there's a worry about invalid URLs and it's possible for multiple embeds to exist, should we perform the `isValidUrl` logic here so we can check all of the available options? i.e.

```js
args.embed = find( post.content_embeds, { src: isValidUrl( src ) } );
```

blowery replied:
`post.content_embeds` has already been vetted and validated, back when it was constructed. The validation on the recieving end in the editor is really more because any old third party could make up a URL that prefills the editor with junk and I wanted to limit the junk to urls.

blowery commented:
@mtias Thanks for the link to the new picker. For sharing, that feels pretty heavy. I was thinking about baking the site picker into the Share button in the Reader, dropping the Twitter and Facebook options, and pushing you direct to a new post in the Editor if you only had one site. So if you have one site, sharing is a one-click thing and if you have more than one, it a click to open and a click to pick. 

How would we do the new Back logic? Does it just use `page.back()` to do the work? Should we do that as part of this PR or another?

mtias commented:
Sounds good, yes, an inline popover/dropdown would work great since we control both ends. It's more for vanilla press-this that the selector would be needed.

> How would we do the new Back logic? Does it just use page.back() to do the work? Should we do that as part of this PR or another?

However you prefer. Probably best to do separately.

blowery commented:
@alisterscott @jancavan Thanks for checking it out! It's definitely early days on this flow. 

Totally agree that we need to handle site selection before dumping folks into the editor. I think we can use the site picker instead of the wp / twitter / fb popup when clicking share and just send 'em over if they only have one site.

The busted embeds are concerning. We're scraping the URL out of the converted iframe and then relying on a shortcode eval to turn it back into an iframe in the editor. @aduth is there a better way to do that? Should we make up an actual iframe instead?

As for the link, I can add the site name to the inbound link. Also need to determine how to handle empty titles. Do we want a separate link to the site, or just one link to the source with the site name as part of the link text? That is [A Collection of Curiously Impossible Dogs on b19y](https://b19y.com/2016/02/24/a-collection-of-curiously-impossible-dogs/) vs [A Collection of Curiously Impossible Dogs](https://b19y.com/2016/02/24/a-collection-of-curiously-impossible-dogs/) on [b19y](https://b19y.com)

aduth commented:
>The busted embeds are concerning. We're scraping the URL out of the converted iframe and then relying on a shortcode eval to turn it back into an iframe in the editor. @aduth is there a better way to do that? Should we make up an actual iframe instead?

The problem is that the URLs in @alisterscott 's examples aren't valid oEmbed links. For example, this link won't embed anywhere:

`https://videopress.com/embed/s3JwoX8z?hd=0&autoPlay=0&permalink=0&loop=0`

But this equivalent link would if pasted in the editor:

`https://videopress.com/v/s3JwoX8z`

To my knowledge, there's not an easy way for us to "reverse oEmbed", i.e. take parts of the embed result and tie it back to an equivalent embeddable URL.

We can't/won't save `iframe`s in the post content. While an oEmbed link appears rich in the editor and on the front-end, the true content saved is the plaintext link.

blowery commented:
@mtias I started digging through how to use SitesPopover / SiteSelector to drive the share popover. It appears that I'll need to add a new prop to pass params along in the href that the Site component ends up using in the list of sites. Does that jive with what you'd expect?

blowery commented:
> To my knowledge, there's not an easy way for us to "reverse oEmbed", i.e. take parts of the embed result and tie it back to an equivalent embeddable URL.

Well that's a bummer. Maybe we need a map of reversals for the common ones?

aduth commented:
>Well that's a bummer. Maybe we need a map of reversals for the common ones?

Brief lapse of memory, but we do have a mapping of _some_ popular embed code back to an equivalent shortcode ("shortcode reversal"). This takes place when a user saves the post, and is intended for services which offer their own embed code that a user is likely to try to paste into their post directly.

Kinda hacky, limited, and hard to rely on. I guess the ideal would be to insert the raw iframe markup in the editor, though I'm not sure how that would display (easy enough to test).

ryanboren commented:
While testing the branch, I noticed on a post detail screen in the reader that I wasn't following the site. That was odd, since I visited the detail view from my Followed Sites  stream. I clicked the follow button and saw this:

![screen shot 2016-02-25 at 2 39 32 pm](https://cloud.githubusercontent.com/assets/2047747/13333923/f368b7d2-dbce-11e5-895f-6c9fdfe27af6.png)

After that, I tried sharing. Click share icon, click WordPress, nothing happens. The editor does not come up. I try a few times with the same result. I give up, clicking browser back history and see this:

![screen shot 2016-02-25 at 2 40 09 pm](https://cloud.githubusercontent.com/assets/2047747/13333971/240f6db8-dbcf-11e5-85da-1b92747dec5c.png)

Peek-a-boo layers.

I tried reproducing with and without first triggering the follow error and couldn't reproduce again.

I see go back flow is being discussed with a s/Posts/Back/ proposal for the editor. Go back flow can make or break this. I want to land right back in the reader stream where I left.

blowery commented:
@ryanboren That should be fixed with #3418, which this PR doesn't have yet. I'll rebase it and pick it up. If you can still repro the bad follow button behavior, I'd love to hear about it.

blowery commented:
Rebased and force pushed

blowery commented:
Had to do a bunch of surgery to get the SitePopover to work in post cards. Might want to break bits of it out to other PRs perhaps.

mtias commented:
> Does that jive with what you'd expect?

Makes sense. Why do you need extra props? Are we stripping the params for the href value?

blowery commented:
rebased and forced. time to squash. getting big.

blowery commented:
OK, I think this is really for hard review. 

Last couple things I really need feedback on:
@aduth @mtias https://github.com/Automattic/wp-calypso/pull/3544/commits/c7fefe34f3bf5cff24091e8c59b970b2174f5e9b changes how we determine if a post is empty in ground control. Without that change, you have to make a meaningless change to the shared post to get it save-able.

I hacked in a Back button if we have a history in `page`. This way, you see ←Posts if you load the page fresh, but any other entry point will get you a ←Back button to go back to where you came from. I kinda like it, as I often step into the editor from places other than the posts page. But not sure it's the best.

aduth commented on npm-shrinkwrap.json:
```diff
@@ -8564,6 +8564,12 @@
     "blob": {
       "version": "0.0.4"
+    },
+    "ms": {
```
Should this have been committed in a previous pull request? Not seeing how this is related to the dependency added here.

blowery replied:
hmmm... the shrinkwrap has conflicted everytime i rebased this. It's possible this slipped in.

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -26,7 +26,7 @@ const Card = require( 'components/card' ),
 
 function isPostEmpty( props ) {
-	return ( props.isNew && ! props.isDirty ) || ! props.hasContent;
+	return ! props.hasContent || ( ! props.isNew && ! props.isDirty );
```
I'm guessing the original logic was such that if the post is new and hasn't been modified, it must be empty? Do we even need to check this? Can't we test only against `hasContent` ?

blowery replied:
Yeah, this is / was used to determine when to light up the publish and preview buttons. The logic is there is pretty convoluted, but covered by tests, and I didn't want to mess with it too much.

blowery replied:
Hrm. new logic still isn't right. Share a post, publish is available. Then auto-save kicks in and publish goes away.

aduth commented on client/post-editor/editor-sidebar/header.jsx:
```diff
@@ -22,19 +24,32 @@ import DraftsButton from 'post-editor/drafts-button';
-		case 'post': allPostsLabel = translate( 'Posts' ); break;
-		case 'page': allPostsLabel = translate( 'Pages' ); break;
-		default: allPostsLabel = get( type, 'labels.menu_name' ) || translate( 'Loading…' );
+	const useBackButton = page.len > 0;
```
If I follow correctly, this changes the behavior such that if the user arrives at the editor from any page, the header displays a "Back" button instead of "Posts"? I'm on the fence with this one, but I'm not strictly opposed to it outside of the concern that it breaks existing expectations. Care to weigh in, @mtias or @folletto ?

blowery replied:
Correct, that's the new behavior. I'm not in love with it, but I don't hate it either. Ideally we'd be able to contextualize the label to what you're heading back to, but I couldn't figure out how to weave that into the information available to the sidebar header.

aduth commented on client/post-editor/editor-sidebar/header.jsx:
```diff
@@ -22,19 +24,32 @@ import DraftsButton from 'post-editor/drafts-button';
+			default: closeLabel = get( type, 'labels.menu_name' ) || translate( 'Loading…' );
+		}
 	}
+	const closeButtonAction = useBackButton ? goBack : noop;
```
Any reason not to use `page.back` ?

https://github.com/visionmedia/page.js/blob/aa60550/index.js#L205-L229

blowery replied:
nope! i'll switch it.

aduth commented on client/post-editor/editor-sidebar/header.jsx:
```diff
@@ -85,7 +101,8 @@ export default connect(
 			type: getPostType( state, siteId, typeSlug ),
-			showDrafts: isEditorDraftsVisible( state )
+			showDrafts: isEditorDraftsVisible( state ),
+			useBackButton: state
```
This should be removed?

aduth commented on client/reader/share/index.jsx:
```diff
@@ -42,30 +45,32 @@ var actionMap = {
+	args.title = `${ post.title } — ${ post.site_name }`;
+	args.text = post.excerpt;
+	args.url = post.URL;
+	args.backto = window.location.href;
```
Is this being used?

blowery replied:
nope, will remove.

aduth commented on client/reader/share/index.jsx:
```diff
@@ -73,18 +78,52 @@ var ReaderShare = React.createClass( {
 
+	componentWillUnmount() {
+		if ( this._closeHandle ) {
+			clearTimeout( this._closeHandle );
```
Aside: I kinda wish `defer` returned a `cancel` function similar to `debounce` and `throttle`. As-is, the overlap with `setTimeout` is pretty confusing (i.e. why use `defer` instead of `setTimeout`, if cancelling is only possible via `clearTimeout`?)

aduth commented on client/reader/share/style.scss:
```diff
@@ -99,3 +99,7 @@
 }
+
+.sites-popover.is-reader .site-selector__sites {
+		max-height: 25vh;
```
Tabbed one too far.

blowery commented on client/components/popover/index.jsx:
```diff
@@ -90,25 +91,29 @@ var Popover = React.createClass( {
+						);
+					}
+
+					this._tip.position( this.props.position, { auto: true } );
```
This needs testing still. @mtias thought it might cause lots of re-renders.

shaunandrews commented:
This is a little awkward for users with one site:

![screen shot 2016-03-23 at 11 16 24 am](https://cloud.githubusercontent.com/assets/191598/13989948/d59f9688-f0e8-11e5-9091-dcaffbaf21c5.png)

@blowery mentioned pushing people straight to the editor in this case, but I'm not sure thats a great solution either — at least the site selector doesn't pull you totally out of the flow. Any ideas on a solution here? I'm sure its crazy to say that we could/should load the editor as a modal?

--

The editor sidebar is doing this weird hide/show dance in Safari:

![screen shot 2016-03-23 at 11 20 27 am](https://cloud.githubusercontent.com/assets/191598/13990091/5c07a6a2-f0e9-11e5-8a20-61ba4f08a8e5.png)

It literally flashed in and and out a dozen times over the course of a few seconds, then just blipped out. Resizing the window fixed it. Fun. But then it happened again after I published:

![screen shot 2016-03-23 at 11 22 37 am](https://cloud.githubusercontent.com/assets/191598/13990123/8e542464-f0e9-11e5-9651-c1cbda4168df.png)

Resizing again fixed it.

--

The transition between the Editor and the Full Post view — in both directions — is a little wonky. Maybe we can ditch the transform on the full post when sharing?

folletto commented:
> Any ideas on a solution here? I'm sure its crazy to say that we could/should load the editor as a modal?

I agree with you. This imho is the ideal scenario, even if it probably the effort has to be done separately and then looped back here?

For reference this was the latest iteration fo the mockup we discussed, in a different context, but it's theoretically ready to work on the Reader too if we add the picker as above (and probably some tweaks to the top bar... like the site visible on the left, and drafts as a third button?).

![customizer-editor-i1](https://cloud.githubusercontent.com/assets/4389/13990334/df4e980a-f10b-11e5-8675-8ea51b64220d.png)

blowery commented:
> Maybe we can ditch the transform on the full post when sharing?

I think that means ditching the transition all the time. We don't have a good way to make that contextual.

blowery commented:
> The editor sidebar is doing this weird hide/show dance in Safari:

Only on this branch, or on master too?

blowery commented:
> @blowery mentioned pushing people straight to the editor in this case, but I'm not sure thats a great solution either — at least the site selector doesn't pull you totally out of the flow. Any ideas on a solution here? I'm sure its crazy to say that we could/should load the editor as a modal?

I actually tried just pushing them into the editor and it's a really bad flow, especially when you accidentally "click" the share button when scrolling through a list on a phone. I think this _has_ to be a two-click thing. It's just too disruptive otherwise.

The modal fixes a bit of it, but I think it would still be overwhelming, especially on a phone. This is an action you have to take with intent to leave your reading flow, so I don't think it's too crazy to make it an effectively "confirmed" action.

folletto commented:
> The modal fixes a bit of it, but I think it would still be overwhelming, especially on a phone. This is an action you have to take with intent to leave your reading flow, so I don't think it's too crazy to make it an effectively "confirmed" action.

I'd highlight that the user would still have to pick the target site before loading the editor, so that UI will avoid the "direct jump" effect. Referring to the mock above, on single site we can just make a UI that has a title at the top like "Share on..." and the site badge. So works on both 1 item and multiple items. :)

blowery commented:
> Referring to the mock above, on single site we can just make a UI that has a title at the top like "Share on..." and the site badge.

I don't follow. The mock above isn't the Reader or a reader stream. Are you saying the Share button in stream should be "Share on $your site" (which will get really repetitive) and open a modal editor directly? Or are you happy with the current picker solution on this branch?

folletto commented:
Ah sorry. I was referring to the reader one, [this](https://cloud.githubusercontent.com/assets/191598/13989948/d59f9688-f0e8-11e5-9091-dcaffbaf21c5.png) one mentioned above. I think that even with just one site, by adding at the top of that popup "Share in..."  or similar text would remove a bit of the awkwardess for one-site users. I consider this as a stop-gap solution, but should be fine. :)

jancavan commented:
>  on single site we can just make a UI that has a title at the top like "Share on..." 

I agree. Probably something like this (quick mockup) to make it more obvious that you're sharing this post on your site. Should just keep the same heading though even with multiple sites.

![share-on](https://cloud.githubusercontent.com/assets/4924246/13994853/f25cc23c-f0e3-11e5-8dd8-13ff81ababd0.jpg)

The popover also doesn't stay attached to the card when scrolling down.

folletto commented:
That! :+1:

blowery commented:
https://cldup.com/uISXiqkAajc/j26KxP.png

blowery commented:
Heh, I'll update to match @jancavan 

Do we want this on the multisite picker too? Above or below the search box?

jancavan commented:
In https://github.com/Automattic/wp-calypso/pull/3544/commits/60383b8adf9965ccc6c4fcfff7288d9e64d81ab5, updated the popover heading styles:

Multi-site:
![share-1](https://cloud.githubusercontent.com/assets/4924246/13996257/22116d56-f0ea-11e5-8c21-2e627628e818.jpg)

Single site:
![screenshot 2016-03-23 11 23 12](https://cloud.githubusercontent.com/assets/4924246/13996245/16f5a14e-f0ea-11e5-812b-dc0f2c3b114a.png)

folletto commented:
:+1: design wise. :)

Aside: shouldn't "Share" be blue when clicked? :)

blowery commented:
Broken on iOS. Doesn't prefill the post for some reason. Looks like it's killing the URL being used and trimming off the querystring?

blowery commented:
> Broken on iOS. Doesn't prefill the post for some reason. Looks like it's killing the URL being used and trimming off the querystring?

fixed that by using onSiteSelect and exposing that through the site popover. The problem was that the SiteSelector was calling `page` if the event wasn't a `mouseup`, and the event on a tablet is a `touchend`. Calling into `page` myself first works, but ...

I can tear out all of the querystring stuff I added now too, since I don't need it any more.

shaunandrews commented:
That little heading helps a ton.

@folletto's comp above is exactly what I had in mind — popping the editor "over top" of the Reader in a modal. It would really help with the flow, keeping you in the context of the stream/post, while making it all feel more fluid and fast.

blowery commented:
> comp above is exactly what I had in mind — popping the editor "over top" of the Reader in a modal. It would really help with the flow, keeping you in the context of the stream/post, while making it all feel more fluid and fast.

Yeah, I'd love to get there, but I think that's out of scope for this PR.

blowery commented:
rebased, squashed down, forced again

mtias commented:
> I can tear out all of the querystring stuff I added now too, since I don't need it any more.

Nice :)

blowery commented:
Worked up https://github.com/component/tip/pull/56 to address the crazy positioning problems we were seeing with popovers. Seems to work quite nicely now.

blowery commented:
@mtias @folletto any objections to the editor changes I made here? Mostly just the back handling.

folletto commented:
> any objections to the editor changes I made here? Mostly just the back handling.

For some reason the first time I click on share after a hard refresh I still get "← POSTS". Bug?

Otherwise seems good. Back there seems a good action for the Reader share.

—

Also another detail: the popup doesn't close once I open it, and clicking anywhere else just triggers that action. It should just close the popup instead (regardless of where one's click, with no triggering of other actions).

blowery commented:
> Also another detail: the popup doesn't close once I open it, and clicking anywhere else just triggers that action. It should just close the popup instead (regardless of where one's click, with no triggering of other actions).

Hrm. That _might_ be fixed with https://github.com/component/tip/pull/56, which hasn't landed yet or been picked up here. But I sort of doubt it.

blowery commented:
> Also another detail: the popup doesn't close once I open it, and clicking anywhere else just triggers that action. It should just close the popup instead (regardless of where one's click, with no triggering of other actions).

@folletto Should be fixed with https://github.com/Automattic/wp-calypso/pull/3544/commits/2d2a369f717ead3ad89a5c6e2cf3701efc5f8ab4. Give it a shot!

folletto commented:
Confirmed :+1:

blowery commented:
I think this is set, but I'm about to step away for the evening. Plan on merging it tomorrow, AM, EDT.
